### ISSUE TITLE
[OJS] Session problems in multijournal installations with custom base_url's

### ISSUE THREAD

**Issue by ajnyga** [pkp/pkp-lib#4595]
Hi, this applies to all latest versions of OJS, because I think the code for sessions has been the same for a while.

Scenario:

User is logged in from the site level at `site.com`. A new session with domain `.site.com` is created.

User moves to a subdomain journal (custom base_url). The user no longer appears to be logged in and a second session is created with domain `.journalname.site.com`. 

User logs in from the journal and can enter the dashboard successfully.

User logs out. 

The session `.journalname.site.com` is updated to reflect that the user has logged out. The session with domain `.site.com` still shows that the user is logged in.

Problem: **if the user now tries to log in from the journal she can not anymore**. After logging in the session is updated to reflect that she has logged in, but OJS does not allow her to access the dashboard. Also the user navigation menu shows the login and register links.

What I am wondering is that do we need to initiate new cookies for each subdomain, or could we just always use the main domain in the cookie? For example `.site.com`.

**Comment by ajnyga** [comment]
@ctgraham would really like to hear your view on this, I think you have dealt with similar issues before.

@asmecher my suggestion is that we allow a config override here: https://github.com/pkp/pkp-lib/blob/master/classes/session/SessionManager.inc.php#L41

The problem with `getServerHost` seems to be that it returns the main domain `site.com` on the site level but returns the journal subdomain `name.site.com` on the context level. So you always end up with two cookies with two domains. 

I believe that this section of code here https://github.com/pkp/pkp-lib/blob/master/classes/session/SessionManager.inc.php#L65-L71 should deal with that issue? But it does not. What comes to mind is that we upgraded to PHP7 recently and these problems started to show up. Could there be something in the sessions management code that changed from 5->7?

In any case, I hardcoded the `site.com` value here https://github.com/pkp/pkp-lib/blob/master/classes/session/SessionManager.inc.php#L41 and it works fine now. 
When I go to the site level I get a session and stay logged in when I move to the subdomain journal. Also, if I first visit a subdomain journal with no existing sessions I still get just one cookie and stay logged in if I move to the site level. I can also log out from either of those places.

So do you see any problems arising if we would allow site admins to define an override cookie domain value from config.inc.php.

Would this work with cases where each journal has it's own domain altogether?

**Comment by ctgraham** [comment]
My earlier investigation of this was largely based on divergent behavior of Internet Explorer:
http://wayback.archive-it.org/7100/20180509142027/https://pkp.sfu.ca/bugzilla/show_bug.cgi?id=8843

My original proposal suggested propagating up a session to the parent domain, if sessions on both the current domain and parent domain were found to belong to the same OJS installation.

I wasn't able to reproduce the issue you describe in my development site with the following configuration:
* Chrome 72.0.3626.121 
* Apache/2.2.15
* PHP 5.6.40
* OJS 3.1.1

Perhaps this is a change from PHP 5 to PHP 7, as you suggested?  I'm in the process of building a PHP7 dev environment for OJS 3.1.2, but it is unlikely I'll be able to report on results this week due to holiday and vacation time.

**Comment by ajnyga** [comment]
I have a feeling that PHP7 could be involved because I do not think these problems occured earlier and we have been using subdomains for 3 years now.

**Comment by ajnyga** [comment]
If you log in on the site level and move on to a subdomain context, do you get two session cookies?

**Comment by ctgraham** [comment]
Not in my testing as above.  Each domain as a single distinct cookie.  What browser are you testing with?

**Comment by ajnyga** [comment]
Ok, so just to be clear, my problem seems to be happening because each domain get a distinct cookie.

I log in in the site level > I get `site.com` cookie.
I move to a subdomain journal > I get new `journalname.site.com` cookie and have login again
I log out in from the context > the `site.com` cookie stays logged in and the `journalname.site.com` is logged out, but something is then preventing me from logging in again to the context.

I figured that the best solution would be if the `site.com` cookie would work also in `journalname.site.com` so that there would be just one cookie all the time?

**Comment by ctgraham** [comment]
> the best solution would be if the site.com cookie would work also in journalname.site.com so that there would be just one cookie all the time?

I concur.  This was the intent of my PR from 2014, though it appears it is not actually working that way, even in PHP5.

**Comment by ajnyga** [comment]
do you see any potential problems in allowing a config override that would just set a certain domain?
Just thinking how it all would work if some journal has a completely own domain.

**Comment by ctgraham** [comment]
I generally take a very negative view of variables in config.inc.php.  We ought to be able to handle this appropriately directly within the session management code.  If we can't, however, the use of a config.inc.php variable might be a solution.  I really strongly lean toward fixing the code.  The config.inc.php override would be problematic if any `base_url[]` did not exist in that "forced" cookie domain.

**Comment by ajnyga** [comment]
The reason why I think a config is the only solution is because the main site may be installed on a subdomain address itself. 

If you have  an organisation at`organisation.com` and they have OJS at `ojs.organisation.com` and subdomain journal at `journal1.ojs.organisation.com`. How do we know that we want to use `ojs.organisation.com` in the cookie domain? Or do you think we could just use `organisation.com`?

My idea was that the setting would be voluntary. Just something that would override this https://github.com/pkp/pkp-lib/blob/master/classes/session/SessionManager.inc.php#L41 if it was defined. Otherwise we would just use getServerHost.

**Comment by ctgraham** [comment]
I don't think we want to send the cookie to all of `organisation.com` for security purposes.

If access starts at `ojs.organisation.com`, and if the cookie for `ojs.organisation.com` is sent to `journal.ojs.organisation.com`, it should be trivial for the code to use the parent's cookie.

If access starts at `journal.ojs.organisation.com`, an additional login will be required if accessing `ojs.organisation.com`, but the next access to `journal.ojs.organisation.com` can delete the `journal.ojs.organisation.com` cookie in favor of the `ojs.organisation.com` cookie.

**Comment by ajnyga** [comment]
hmm, ok. I will see if I can turn that into a pr

**Comment by ajnyga** [comment]
Your suggestion can be reached by adding 

```
		if (isset($this->userSession))
			$domain = $this->userSession->getDomain();
		else
			$domain = ini_get('session.cookie_domain');
```

here: https://github.com/pkp/pkp-lib/blob/master/classes/session/SessionManager.inc.php#L218

But I still think that allowing a domain to be set from config.inc.php would be a cleaner solution.

**Comment by ajnyga** [comment]
The reason why you end up with multiple cookies is that setcookie here: https://github.com/pkp/pkp-lib/blob/master/classes/session/SessionManager.inc.php#L227

is called with different domain names. Namely the main domain and the subdomains when you go to a context. 

So maybe alternatively we could always delete the old cookie before setting the new one?

**Comment by ajnyga** [comment]
@ctgraham not seeing many problems with that solution https://github.com/pkp/pkp-lib/issues/4595#issuecomment-472519360

If you access the site from the site level, you end up with a single cookie. If you access from the subdomain journal, you end up with two and have to log in again on site level. Just like you described. But compared with the original situation that is not that bad.

Also I believe that this solution would work the best with installations that have separate domains for journals? I think that my original solution with a fixed cookie domain name would not work at all.

@asmecher would really appreciate your comments here as well. This has become a fairly big issue at least for us. But of course understand that you have other things you are busy with as well.

**Comment by asmecher** [comment]
Hi all -- I've been following the conversation but haven't had much to add as our hosting team doesn't do much multi-journal hosting with different domains. I suspect the two of you have better use cases/test environments/etc. Generally this code has been stable and widely deployed, including in rewrite scenarios, so I do suspect something changed on the underlying platform e.g. with PHP7 support.

The proposal to add a `config.inc.php` override on a sensible but sometimes imperfect default generally serves us well (your proposal at https://github.com/pkp/pkp-lib/issues/4595#issuecomment-472520128, @ajnyga). To support this most flexibly, I would think we would need to support the option of a series of overrides, à la `base_url[...]`, for each domain.) I wonder whether this would be useful for other exotic rewrite scenarios where the actual domain is not available to pkp-lib. In any case, it's unobtrusive for those who don't want it.

At the same time, I can't think of any downsides to @ctgraham's proposal either. I don't think it completely solves the problem (the first call to `setcookie` may be using too specific a domain, and subsequent requests may not even send the cookie as a result) but it does mean that at least we're being wrong consistently. The domain information is stored server-side so there's no risk of opening anything up accidentally.

The two would pair together to provide complete coverage, as I see it.

**Comment by ajnyga** [comment]
@ctgraham we are getting this issue more frequently nowadays.

Which of the two solutions do you prefer. I have to move this to my high priority list and fix it asap.

**Comment by ajnyga** [comment]
lib: https://github.com/pkp/pkp-lib/pull/6305
ojs: https://github.com/pkp/ojs/pull/2891 (will add OPS/OMP when the concept is ok)

@asmecher @ctgraham

**Comment by ajnyga** [comment]
This definitely needs double and triple checking so that my pr's do not break anything!

**Comment by ajnyga** [comment]
also @NateWr if you have time to check this, would probably be a good idea. The original problem described here https://github.com/pkp/pkp-lib/issues/4595#issue-420369909

**Comment by ajnyga** [comment]
Just want to add that we have a few scenarios we should support:

1. Default: single domain name is used. This works fine with the current code.
2. Subdomains: some or all journals use subdomains and the site uses the parent domain
3. Own domains: there is a parent site and a single installation, but some or all journals use their own domains. 
4. A combination of the three above

Option 3 of course involves other problems, like the shared user database. So one could argue that we should only support 3, if other necessary changes are considered as well.

**Comment by NateWr** [comment]
The URL rewriting is not something I know much about! But what about cases where each journal is on its own domain and we don't want to share logins? I think this is a requirement for PKP|PS, which wants to be able to run several journals through one instance with no public relationship between journals.

**Comment by ajnyga** [comment]
I have been trying to figure this.

It seems that the current OJS master code works fine when journals use their own domains. I mean for example site.com journal1.com journal2.com etc. Each domain gets their own cookie/session and the bug desribed above does not happen.

This only happens with subdomains. Here is the best description I can create:

Login to http://site.com/publicknowledge
Cookie 461e747c16897ed2c64a918dbeb542f5 .site.com created
Session 461e747c16897ed2c64a918dbeb542f5 .site.com created
Log out

Go to http://journal2.site.com
Cookie 461e747c16897ed2c64a918dbeb542f5 .site.com available, 
the code does not create a new cookie/session for journal2.site.com at this point

Log in to http://journal2.site.com
Cookie 8294fea74eb92fd1248891ab75a56c5f .journal2.site.com created
Session 8294fea74eb92fd1248891ab75a56c5f .journal2.site.com created BUT user_id  is set to NULL
Session 461e747c16897ed2c64a918dbeb542f5 .site.com session user_id has changed to 1
No error, but the user is not logged in!

**Basically the login for journal2.site.com is trying to use the .site.com session instead of using the new session for .journal2.site.com**

**Comment by asmecher** [comment]
But your PRs offer a solution for that situation, don't they? (https://github.com/pkp/pkp-lib/issues/4595#issuecomment-718379310)

**Comment by ajnyga** [comment]
The config.inc.php setting offers a solutions if the site _only_ uses subdomains. It will not work if the site has both subdomains and journals with their own domains. Current master works fine with everthing but subdomains.

**Comment by ajnyga** [comment]
I just do not understand where the last step th new session gets domain site.com, because I have been trying to debug this and I can not find a place in the code where site.com is returned instead of journal2.site.com.

**Comment by asmecher** [comment]
@ctgraham, Antti-Jussi's latest PR (https://github.com/pkp/pkp-lib/pull/6315) looks pretty clear to me, but I thought I'd ping you for an opinion too...

**Comment by pverberne** [comment]
We have similar problem: I host 4 journals in a subdomain in 1 installation of OJS.
config.inc.php:

base_url = "https://openjournals.maastrichtuniversity.nl"
base_url[index] = "https://openjournals.maastrichtuniversity.nl"
base_url[mjla] = "https://openjournals.maastrichtuniversity.nl/mjla"
base_url[msjpn] = "https://openjournals.maastrichtuniversity.nl/msjpn"
base_url[SustainabilityStudies] = "https://openjournals.maastrichtuniversity.nl/SustainabilityStudies"
base_url[SL] = "https://www.sustainabilityletters.net/SL"
As you see, The 5th journal has it's own domain. The frontend is working just fine, but you access the backend of o "https://www.sustainabilityletters.net".
